File: packages/synapse-sdk-py/src/synapse_sdk/mcp.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Callable, Protocol


class MCPToolCaller(Protocol):
    def __call__(self, tool_name: str, arguments: dict[str, Any]) -> Any: ...


MCP_CONTEXT_POLICY_PROFILES: dict[str, dict[str, Any]] = {
    "off": {
        "context_policy_mode": "off",
        "description": "Disable context policy filtering and rely on baseline ranking only.",
    },
    "advisory": {
        "context_policy_mode": "advisory",
        "description": "Keep full retrieval set but attach policy/confidence diagnostics.",
    },
    "enforced": {
        "context_policy_mode": "enforced",
        "min_retrieval_confidence": 0.45,
        "min_total_score": 0.20,
        "min_lexical_score": 0.08,
        "min_token_overlap_ratio": 0.15,
        "description": "Filter low-confidence rows for production-safe prompt injection.",
    },
    "strict_enforced": {
        "context_policy_mode": "enforced",
        "min_retrieval_confidence": 0.60,
        "min_total_score": 0.30,
        "min_lexical_score": 0.10,
        "min_token_overlap_ratio": 0.20,
        "description": "Use stricter filtering for high-risk or high-precision workflows.",
    },
}


def _coerce_payload(response: Any) -> Any:
    if isinstance(response, dict):
        if "structuredContent" in response and isinstance(response["structuredContent"], dict):
            return response["structuredContent"]
        if "result" in response and isinstance(response["result"], dict):
            return response["result"]
        return response
    if isinstance(response, list):
        return response
    if hasattr(response, "structuredContent"):
        structured = getattr(response, "structuredContent")
        if isinstance(structured, dict):
            return structured
    if hasattr(response, "content"):
        content = getattr(response, "content")
        if isinstance(content, list):
            for item in content:
                if isinstance(item, dict) and isinstance(item.get("text"), str):
                    try:
                        return json.loads(item["text"])
                    except Exception:
                        continue
    return {"raw": response}


def _normalize_context_policy_mode(value: str | None) -> str:
    normalized = str(value or "").strip().lower()
    if normalized in {"off", "advisory", "enforced"}:
        return normalized
    return "advisory"


def _normalize_context_policy_profile(value: str | None) -> str | None:
    normalized = str(value or "").strip().lower()
    if not normalized:
        return None
    if normalized in MCP_CONTEXT_POLICY_PROFILES:
        return normalized
    allowed = ", ".join(sorted(MCP_CONTEXT_POLICY_PROFILES.keys()))
    raise ValueError(f"unsupported context policy profile `{value}` (allowed: {allowed})")


def _resolve_context_policy_profile(value: str | None) -> dict[str, Any]:
    profile_name = _normalize_context_policy_profile(value)
    if profile_name is None:
        return {}
    profile = MCP_CONTEXT_POLICY_PROFILES.get(profile_name) or {}
    return {
        "context_policy_profile": profile_name,
        "context_policy_mode": profile.get("context_policy_mode"),
        "min_retrieval_confidence": profile.get("min_retrieval_confidence"),
        "min_total_score": profile.get("min_total_score"),
        "min_lexical_score": profile.get("min_lexical_score"),
        "min_token_overlap_ratio": profile.get("min_token_overlap_ratio"),
    }


def list_context_policy_profiles() -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for name in sorted(MCP_CONTEXT_POLICY_PROFILES.keys()):
        profile = dict(MCP_CONTEXT_POLICY_PROFILES.get(name) or {})
        profile["profile"] = name
        out.append(profile)
    return out


def _clamp_float(value: Any, *, minimum: float, maximum: float, default: float) -> float:
    try:
        parsed = float(value)
    except Exception:
        return float(default)
    return max(minimum, min(maximum, parsed))


def _coerce_context_policy_payload(search_payload: dict[str, Any], *, resolved_mode: str) -> dict[str, Any]:
    explainability = search_payload.get("explainability")
    if isinstance(explainability, dict):
        context_policy = explainability.get("context_policy")
        if isinstance(context_policy, dict):
            return context_policy
    payload_policy = search_payload.get("context_policy")
    if isinstance(payload_policy, dict):
        return payload_policy
    return {"mode": resolved_mode}


def _filter_context_results(
    results: list[Any],
    *,
    mode: str,
    min_retrieval_confidence: float | None,
) -> tuple[list[Any], int]:
    if mode != "enforced":
        return list(results), 0
    threshold = (
        _clamp_float(min_retrieval_confidence, minimum=0.0, maximum=1.0, default=0.45)
        if min_retrieval_confidence is not None
        else None
    )
    kept: list[Any] = []
    filtered_out = 0
    for item in results:
        if not isinstance(item, dict):
            kept.append(item)
            continue
        context_policy = item.get("context_policy")
        if isinstance(context_policy, dict) and "eligible" in context_policy:
            if bool(context_policy.get("eligible")):
                kept.append(item)
            else:
                filtered_out += 1
            continue
        confidence_raw = item.get("retrieval_confidence")
        if threshold is not None and isinstance(confidence_raw, (int, float)) and float(confidence_raw) < threshold:
            filtered_out += 1
            continue
        kept.append(item)
    return kept, filtered_out


@dataclass
class MCPContextHelper:
    project_id: str
    call_tool: MCPToolCaller
    default_search_limit: int = 6
    default_fact_limit: int = 20
    default_context_policy_profile: str | None = None
    default_context_policy_mode: str = "advisory"
    default_min_retrieval_confidence: float | None = None
    default_min_total_score: float | None = None
    default_min_lexical_score: float | None = None
    default_min_token_overlap_ratio: float | None = None

    def search_knowledge(
        self,
        query: str,
        *,
        limit: int | None = None,
        filters: dict[str, Any] | None = None,
        context_policy_profile: str | None = None,
        context_policy_mode: str | None = None,
        min_retrieval_confidence: float | None = None,
        min_total_score: float | None = None,
        min_lexical_score: float | None = None,
        min_token_overlap_ratio: float | None = None,
    ) -> dict[str, Any]:
        requested_limit = int(limit if limit is not None else self.default_search_limit)
        payload = {
            "project_id": self.project_id,
            "query": query,
            "limit": max(1, min(100, requested_limit)),
        }
        profile_overrides = _resolve_context_policy_profile(context_policy_profile or self.default_context_policy_profile)
        resolved_mode = _normalize_context_policy_mode(
            context_policy_mode or profile_overrides.get("context_policy_mode") or self.default_context_policy_mode
        )
        if resolved_mode != "advisory":
            payload["context_policy_mode"] = resolved_mode

        resolved_min_confidence = (
            min_retrieval_confidence
            if min_retrieval_confidence is not None
            else (
                profile_overrides.get("min_retrieval_confidence")
                if profile_overrides.get("min_retrieval_confidence") is not None
                else self.default_min_retrieval_confidence
            )
        )
        if resolved_min_confidence is not None:
            payload["min_retrieval_confidence"] = _clamp_float(
                resolved_min_confidence,
                minimum=0.0,
                maximum=1.0,
                default=0.45,
            )
        resolved_min_total = (
            min_total_score
            if min_total_score is not None
            else (
                profile_overrides.get("min_total_score")
                if profile_overrides.get("min_total_score") is not None
                else self.default_min_total_score
            )
        )
        if resolved_min_total is not None:
            payload["min_total_score"] = _clamp_float(
                resolved_min_total,
                minimum=0.0,
                maximum=2.0,
                default=0.20,
            )
        resolved_min_lexical = (
            min_lexical_score
            if min_lexical_score is not None
            else (
                profile_overrides.get("min_lexical_score")
                if profile_overrides.get("min_lexical_score") is not None
                else self.default_min_lexical_score
            )
        )
        if resolved_min_lexical is not None:
            payload["min_lexical_score"] = _clamp_float(
                resolved_min_lexical,
                minimum=0.0,
                maximum=2.0,
                default=0.08,
            )
        resolved_min_overlap = (
            min_token_overlap_ratio
            if min_token_overlap_ratio is not None
            else (
                profile_overrides.get("min_token_overlap_ratio")
                if profile_overrides.get("min_token_overlap_ratio") is not None
                else self.default_min_token_overlap_ratio
            )
        )
        if resolved_min_overlap is not None:
            payload["min_token_overlap_ratio"] = _clamp_float(
                resolved_min_overlap,
                minimum=0.0,
                maximum=1.0,
                default=0.15,
            )

        payload.update(dict(filters or {}))
        response = self.call_tool("search_knowledge", payload)
        normalized = _coerce_payload(response)
        if isinstance(normalized, dict):
            return normalized
        return {"results": normalized}

    def get_entity_facts(
        self,
        entity_key: str,
        *,
        limit: int | None = None,
        category: str | None = None,
        include_non_current: bool = False,
    ) -> dict[str, Any]:
        requested_limit = int(limit if limit is not None else self.default_fact_limit)
        payload = {
            "project_id": self.project_id,
            "entity_key": entity_key,
            "limit": max(1, min(500, requested_limit)),
            "include_non_current": bool(include_non_current),
        }
        if category:
            payload["category"] = category
        response = self.call_tool("get_entity_facts", payload)
        normalized = _coerce_payload(response)
        if isinstance(normalized, dict):
            return normalized
        return {"facts": normalized}

    def get_recent_changes(self, *, limit: int = 20, since_hours: int = 168) -> dict[str, Any]:
        payload = {
            "project_id": self.project_id,
            "limit": max(1, min(200, int(limit))),
            "since_hours": max(1, min(24 * 90, int(since_hours))),
        }
        response = self.call_tool("get_recent_changes", payload)
        normalized = _coerce_payload(response)
        if isinstance(normalized, dict):
            return normalized
        return {"changes": normalized}

    def explain_conflicts(
        self,
        *,
        limit: int = 20,
        resolution_status: str = "open",
        entity_key: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "project_id": self.project_id,
            "limit": max(1, min(200, int(limit))),
            "resolution_status": resolution_status,
        }
        if entity_key:
            payload["entity_key"] = entity_key
        response = self.call_tool("explain_conflicts", payload)
        normalized = _coerce_payload(response)
        if isinstance(normalized, dict):
            return normalized
        return {"conflicts": normalized}

    def build_context(
        self,
        *,
        query: str,
        entity_key: str | None = None,
        include_recent_changes: bool = False,
        recent_since_hours: int = 24,
        context_policy_profile: str | None = None,
        context_policy_mode: str = "enforced",
        min_retrieval_confidence: float | None = None,
        min_total_score: float | None = None,
        min_lexical_score: float | None = None,
        min_token_overlap_ratio: float | None = None,
    ) -> dict[str, Any]:
        resolved_mode = _normalize_context_policy_mode(context_policy_mode)
        search_payload = self.search_knowledge(
            query=query,
            context_policy_profile=context_policy_profile,
            context_policy_mode=resolved_mode,
            min_retrieval_confidence=min_retrieval_confidence,
            min_total_score=min_total_score,
            min_lexical_score=min_lexical_score,
            min_token_overlap_ratio=min_token_overlap_ratio,
        )
        raw_results = list(search_payload.get("results", []))
        search_results, filtered_out_local = _filter_context_results(
            raw_results,
            mode=resolved_mode,
            min_retrieval_confidence=min_retrieval_confidence,
        )
        policy_filtered_out = int(search_payload.get("policy_filtered_out") or 0)
        if policy_filtered_out <= 0 and filtered_out_local > 0:
            policy_filtered_out = filtered_out_local
        context: dict[str, Any] = {
            "query": query,
            "search_results": search_results,
            "revision": search_payload.get("revision"),
            "policy_filtered_out": policy_filtered_out,
            "context_policy": _coerce_context_policy_payload(search_payload, resolved_mode=resolved_mode),
        }
        if entity_key:
            facts_payload = self.get_entity_facts(entity_key)
            context["entity_key"] = entity_key
            context["entity_facts"] = list(facts_payload.get("facts", []))
            context["facts_revision"] = facts_payload.get("revision")
        else:
            context["entity_facts"] = []

        if include_recent_changes:
            changes_payload = self.get_recent_changes(since_hours=recent_since_hours)
            context["recent_changes"] = list(changes_payload.get("changes", []))
        else:
            context["recent_changes"] = []
        return context

    def build_context_markdown(
        self,
        *,
        query: str,
        entity_key: str | None = None,
        include_recent_changes: bool = False,
        context_policy_profile: str | None = None,
        max_search_results: int = 5,
        max_entity_facts: int = 8,
        max_recent_changes: int = 4,
    ) -> str:
        context = self.build_context(
            query=query,
            entity_key=entity_key,
            include_recent_changes=include_recent_changes,
            context_policy_profile=context_policy_profile,
        )
        lines: list[str] = ["## Synapse Context Injection"]
        lines.append(f"- query: {query}")
        if entity_key:
            lines.append(f"- entity_key: {entity_key}")

        results = context.get("search_results", [])
        if results:
            lines.append("### Relevant Knowledge")
            for item in results[: max(1, int(max_search_results))]:
                page = item.get("page") if isinstance(item, dict) else {}
                page_slug = page.get("slug") if isinstance(page, dict) else None
                statement = item.get("statement_text") if isinstance(item, dict) else None
                if statement:
                    lines.append(f"- {statement}" + (f" (`{page_slug}`)" if page_slug else ""))

        facts = context.get("entity_facts", [])
        if facts:
            lines.append("### Entity Facts")
            for item in facts[: max(1, int(max_entity_facts))]:
                statement = item.get("statement_text") if isinstance(item, dict) else None
                if statement:
                    lines.append(f"- {statement}")

        changes = context.get("recent_changes", [])
        if changes:
            lines.append("### Recent Changes")
            for item in changes[: max(1, int(max_recent_changes))]:
                action = item.get("action") if isinstance(item, dict) else None
                created_at = item.get("created_at") if isinstance(item, dict) else None
                page = item.get("page") if isinstance(item, dict) else {}
                page_slug = page.get("slug") if isinstance(page, dict) else None
                label = " ".join(part for part in [action, page_slug] if part)
                if label:
                    lines.append(f"- {label}" + (f" ({created_at})" if created_at else ""))

        return "\n".join(lines).strip() + "\n"

    def make_openclaw_search_callback(
        self,
        *,
        default_filters: dict[str, Any] | None = None,
        context_policy_profile: str | None = None,
        context_policy_mode: str | None = None,
        min_retrieval_confidence: float | None = None,
        min_total_score: float | None = None,
        min_lexical_score: float | None = None,
        min_token_overlap_ratio: float | None = None,
    ) -> Callable[..., Any]:
        def _callback(query: str, limit: int = 5, filters: dict[str, Any] | None = None) -> Any:
            merged_filters: dict[str, Any] = {}
            merged_filters.update(default_filters or {})
            merged_filters.update(filters or {})
            payload = self.search_knowledge(
                query=query,
                limit=limit,
                filters=merged_filters,
                context_policy_profile=context_policy_profile,
                context_policy_mode=context_policy_mode,
                min_retrieval_confidence=min_retrieval_confidence,
                min_total_score=min_total_score,
                min_lexical_score=min_lexical_score,
                min_token_overlap_ratio=min_token_overlap_ratio,
            )
            return payload.get("results", payload)

        return _callback

File: packages/synapse-sdk-ts/src/mcp.ts
export type MCPToolCaller = (toolName: string, args: Record<string, unknown>) => Promise<unknown> | unknown;

export type MCPContextPolicyMode = "off" | "advisory" | "enforced";
export type MCPContextPolicyProfileName = "off" | "advisory" | "enforced" | "strict_enforced";

export interface MCPContextPolicyProfile {
  contextPolicyMode: MCPContextPolicyMode;
  minRetrievalConfidence?: number;
  minTotalScore?: number;
  minLexicalScore?: number;
  minTokenOverlapRatio?: number;
  description: string;
}

export const MCP_CONTEXT_POLICY_PROFILES: Record<MCPContextPolicyProfileName, MCPContextPolicyProfile> = {
  off: {
    contextPolicyMode: "off",
    description: "Disable context policy filtering and rely on baseline ranking only."
  },
  advisory: {
    contextPolicyMode: "advisory",
    description: "Keep full retrieval set but attach policy/confidence diagnostics."
  },
  enforced: {
    contextPolicyMode: "enforced",
    minRetrievalConfidence: 0.45,
    minTotalScore: 0.2,
    minLexicalScore: 0.08,
    minTokenOverlapRatio: 0.15,
    description: "Filter low-confidence rows for production-safe prompt injection."
  },
  strict_enforced: {
    contextPolicyMode: "enforced",
    minRetrievalConfidence: 0.6,
    minTotalScore: 0.3,
    minLexicalScore: 0.1,
    minTokenOverlapRatio: 0.2,
    description: "Use stricter filtering for high-risk or high-precision workflows."
  }
};

type ResolvedContextPolicyProfile = {
  contextPolicyProfile?: MCPContextPolicyProfileName;
  contextPolicyMode?: MCPContextPolicyMode;
  minRetrievalConfidence?: number;
  minTotalScore?: number;
  minLexicalScore?: number;
  minTokenOverlapRatio?: number;
};

export function listContextPolicyProfiles(): Array<MCPContextPolicyProfile & { profile: MCPContextPolicyProfileName }> {
  return (Object.keys(MCP_CONTEXT_POLICY_PROFILES) as MCPContextPolicyProfileName[]).map((profile) => ({
    profile,
    ...MCP_CONTEXT_POLICY_PROFILES[profile]
  }));
}

export interface MCPContextHelperOptions {
  defaultSearchLimit?: number;
  defaultFactLimit?: number;
  defaultContextPolicyProfile?: MCPContextPolicyProfileName | string;
  defaultContextPolicyMode?: MCPContextPolicyMode;
  defaultMinRetrievalConfidence?: number;
  defaultMinTotalScore?: number;
  defaultMinLexicalScore?: number;
  defaultMinTokenOverlapRatio?: number;
}

export interface SearchKnowledgeOptions {
  limit?: number;
  filters?: Record<string, unknown>;
  contextPolicyProfile?: MCPContextPolicyProfileName | string;
  contextPolicyMode?: MCPContextPolicyMode;
  minRetrievalConfidence?: number;
  minTotalScore?: number;
  minLexicalScore?: number;
  minTokenOverlapRatio?: number;
}

export interface EntityFactOptions {
  limit?: number;
  category?: string;
  includeNonCurrent?: boolean;
}

export interface BuildContextOptions {
  query: string;
  entityKey?: string;
  includeRecentChanges?: boolean;
  recentSinceHours?: number;
  contextPolicyProfile?: MCPContextPolicyProfileName | string;
  contextPolicyMode?: MCPContextPolicyMode;
  minRetrievalConfidence?: number;
  minTotalScore?: number;
  minLexicalScore?: number;
  minTokenOverlapRatio?: number;
}

export interface BuildContextMarkdownOptions extends BuildContextOptions {
  maxSearchResults?: number;
  maxEntityFacts?: number;
  maxRecentChanges?: number;
}

export class MCPContextHelper {
  private readonly defaultSearchLimit: number;
  private readonly defaultFactLimit: number;
  private readonly defaultContextPolicyProfile?: MCPContextPolicyProfileName;
  private readonly defaultContextPolicyMode: MCPContextPolicyMode;
  private readonly defaultMinRetrievalConfidence?: number;
  private readonly defaultMinTotalScore?: number;
  private readonly defaultMinLexicalScore?: number;
  private readonly defaultMinTokenOverlapRatio?: number;

  constructor(
    private readonly projectId: string,
    private readonly callTool: MCPToolCaller,
    options: MCPContextHelperOptions = {}
  ) {
    this.defaultSearchLimit = normalizeLimit(options.defaultSearchLimit ?? 6, 1, 100, 6);
    this.defaultFactLimit = normalizeLimit(options.defaultFactLimit ?? 20, 1, 500, 20);
    this.defaultContextPolicyProfile = normalizePolicyProfile(options.defaultContextPolicyProfile);
    this.defaultContextPolicyMode = normalizePolicyMode(options.defaultContextPolicyMode);
    this.defaultMinRetrievalConfidence =
      options.defaultMinRetrievalConfidence == null
        ? undefined
        : clampFloat(options.defaultMinRetrievalConfidence, 0, 1, 0.45);
    this.defaultMinTotalScore =
      options.defaultMinTotalScore == null ? undefined : clampFloat(options.defaultMinTotalScore, 0, 2, 0.2);
    this.defaultMinLexicalScore =
      options.defaultMinLexicalScore == null ? undefined : clampFloat(options.defaultMinLexicalScore, 0, 2, 0.08);
    this.defaultMinTokenOverlapRatio =
      options.defaultMinTokenOverlapRatio == null
        ? undefined
        : clampFloat(options.defaultMinTokenOverlapRatio, 0, 1, 0.15);
  }

  async searchKnowledge(query: string, options: SearchKnowledgeOptions = {}): Promise<Record<string, unknown>> {
    const payload: Record<string, unknown> = {
      project_id: this.projectId,
      query,
      limit: normalizeLimit(options.limit ?? this.defaultSearchLimit, 1, 100, this.defaultSearchLimit)
    };
    const profile = resolveContextPolicyProfile(options.contextPolicyProfile ?? this.defaultContextPolicyProfile);
    const mode = normalizePolicyMode(options.contextPolicyMode ?? profile.contextPolicyMode ?? this.defaultContextPolicyMode);
    if (mode !== "advisory") {
      payload.context_policy_mode = mode;
    }
    const minRetrievalConfidence =
      options.minRetrievalConfidence == null
        ? (profile.minRetrievalConfidence ?? this.defaultMinRetrievalConfidence)
        : options.minRetrievalConfidence;
    if (minRetrievalConfidence != null) {
      payload.min_retrieval_confidence = clampFloat(minRetrievalConfidence, 0, 1, 0.45);
    }
    const minTotalScore =
      options.minTotalScore == null ? (profile.minTotalScore ?? this.defaultMinTotalScore) : options.minTotalScore;
    if (minTotalScore != null) {
      payload.min_total_score = clampFloat(minTotalScore, 0, 2, 0.2);
    }
    const minLexicalScore =
      options.minLexicalScore == null
        ? (profile.minLexicalScore ?? this.defaultMinLexicalScore)
        : options.minLexicalScore;
    if (minLexicalScore != null) {
      payload.min_lexical_score = clampFloat(minLexicalScore, 0, 2, 0.08);
    }
    const minTokenOverlapRatio =
      options.minTokenOverlapRatio == null
        ? (profile.minTokenOverlapRatio ?? this.defaultMinTokenOverlapRatio)
        : options.minTokenOverlapRatio;
    if (minTokenOverlapRatio != null) {
      payload.min_token_overlap_ratio = clampFloat(minTokenOverlapRatio, 0, 1, 0.15);
    }
    Object.assign(payload, options.filters ?? {});
    const response = await this.callTool("search_knowledge", payload);
    const normalized = coercePayload(response);
    if (isRecord(normalized)) {
      return normalized;
    }
    return { results: normalized };
  }

  async getEntityFacts(entityKey: string, options: EntityFactOptions = {}): Promise<Record<string, unknown>> {
    const payload: Record<string, unknown> = {
      project_id: this.projectId,
      entity_key: entityKey,
      limit: normalizeLimit(options.limit ?? this.defaultFactLimit, 1, 500, this.defaultFactLimit),
      include_non_current: Boolean(options.includeNonCurrent ?? false)
    };
    if (options.category) {
      payload.category = options.category;
    }
    const response = await this.callTool("get_entity_facts", payload);
    const normalized = coercePayload(response);
    if (isRecord(normalized)) {
      return normalized;
    }
    return { facts: normalized };
  }

  async getRecentChanges(limit = 20, sinceHours = 168): Promise<Record<string, unknown>> {
    const payload = {
      project_id: this.projectId,
      limit: normalizeLimit(limit, 1, 200, 20),
      since_hours: normalizeLimit(sinceHours, 1, 24 * 90, 168)
    };
    const response = await this.callTool("get_recent_changes", payload);
    const normalized = coercePayload(response);
    if (isRecord(normalized)) {
      return normalized;
    }
    return { changes: normalized };
  }

  async explainConflicts(limit = 20, resolutionStatus = "open", entityKey?: string): Promise<Record<string, unknown>> {
    const payload: Record<string, unknown> = {
      project_id: this.projectId,
      limit: normalizeLimit(limit, 1, 200, 20),
      resolution_status: resolutionStatus
    };
    if (entityKey) {
      payload.entity_key = entityKey;
    }
    const response = await this.callTool("explain_conflicts", payload);
    const normalized = coercePayload(response);
    if (isRecord(normalized)) {
      return normalized;
    }
    return { conflicts: normalized };
  }

  async buildContext(options: BuildContextOptions): Promise<Record<string, unknown>> {
    const policyMode = normalizePolicyMode(options.contextPolicyMode ?? "enforced");
    const searchPayload = await this.searchKnowledge(options.query, {
      contextPolicyProfile: options.contextPolicyProfile,
      contextPolicyMode: policyMode,
      minRetrievalConfidence: options.minRetrievalConfidence,
      minTotalScore: options.minTotalScore,
      minLexicalScore: options.minLexicalScore,
      minTokenOverlapRatio: options.minTokenOverlapRatio
    });
    const rawSearchResults = toArray(searchPayload.results);
    const filtered = filterContextResults(rawSearchResults, {
      mode: policyMode,
      minRetrievalConfidence: options.minRetrievalConfidence
    });
    const context: Record<string, unknown> = {
      query: options.query,
      search_results: filtered.results,
      revision: searchPayload.revision ?? null,
      policy_filtered_out:
        typeof searchPayload.policy_filtered_out === "number"
          ? Number(searchPayload.policy_filtered_out)
          : filtered.filteredOut,
      context_policy: coerceContextPolicyPayload(searchPayload, policyMode),
      entity_facts: [],
      recent_changes: []
    };

    if (options.entityKey) {
      const factsPayload = await this.getEntityFacts(options.entityKey);
      context.entity_key = options.entityKey;
      context.entity_facts = toArray(factsPayload.facts);
      context.facts_revision = factsPayload.revision ?? null;
    }

    if (options.includeRecentChanges) {
      const changesPayload = await this.getRecentChanges(20, options.recentSinceHours ?? 24);
      context.recent_changes = toArray(changesPayload.changes);
    }
    return context;
  }

  async buildContextMarkdown(options: BuildContextMarkdownOptions): Promise<string> {
    const maxSearchResults = normalizeLimit(options.maxSearchResults ?? 5, 1, 50, 5);
    const maxEntityFacts = normalizeLimit(options.maxEntityFacts ?? 8, 1, 50, 8);
    const maxRecentChanges = normalizeLimit(options.maxRecentChanges ?? 4, 1, 50, 4);
    const context = await this.buildContext(options);

    const lines: string[] = ["## Synapse Context Injection", `- query: ${options.query}`];
    if (options.entityKey) {
      lines.push(`- entity_key: ${options.entityKey}`);
    }

    const searchResults = toArray(context.search_results);
    if (searchResults.length > 0) {
      lines.push("### Relevant Knowledge");
      for (const item of searchResults.slice(0, maxSearchResults)) {
        if (!isRecord(item)) continue;
        const statement = typeof item.statement_text === "string" ? item.statement_text : "";
        if (!statement) continue;
        const page = isRecord(item.page) ? item.page : {};
        const pageSlug = typeof page.slug === "string" ? page.slug : "";
        lines.push(pageSlug ? `- ${statement} (\`${pageSlug}\`)` : `- ${statement}`);
      }
    }

    const facts = toArray(context.entity_facts);
    if (facts.length > 0) {
      lines.push("### Entity Facts");
      for (const item of facts.slice(0, maxEntityFacts)) {
        if (!isRecord(item)) continue;
        const statement = typeof item.statement_text === "string" ? item.statement_text : "";
        if (statement) {
          lines.push(`- ${statement}`);
        }
      }
    }

    const recentChanges = toArray(context.recent_changes);
    if (recentChanges.length > 0) {
      lines.push("### Recent Changes");
      for (const item of recentChanges.slice(0, maxRecentChanges)) {
        if (!isRecord(item)) continue;
        const action = typeof item.action === "string" ? item.action : "";
        const createdAt = typeof item.created_at === "string" ? item.created_at : "";
        const page = isRecord(item.page) ? item.page : {};
        const pageSlug = typeof page.slug === "string" ? page.slug : "";
        const label = [action, pageSlug].filter(Boolean).join(" ");
        if (label) {
          lines.push(createdAt ? `- ${label} (${createdAt})` : `- ${label}`);
        }
      }
    }
    return `${lines.join("\n").trim()}\n`;
  }

  makeOpenClawSearchCallback(
    defaultFilters: Record<string, unknown> = {},
    options: {
      contextPolicyProfile?: MCPContextPolicyProfileName | string;
      contextPolicyMode?: MCPContextPolicyMode;
      minRetrievalConfidence?: number;
      minTotalScore?: number;
      minLexicalScore?: number;
      minTokenOverlapRatio?: number;
    } = {}
  ) {
    return async (query: string, limit = 5, filters: Record<string, unknown> = {}): Promise<unknown> => {
      const payload = await this.searchKnowledge(query, {
        limit,
        filters: { ...defaultFilters, ...filters },
        contextPolicyProfile: options.contextPolicyProfile,
        contextPolicyMode: options.contextPolicyMode,
        minRetrievalConfidence: options.minRetrievalConfidence,
        minTotalScore: options.minTotalScore,
        minLexicalScore: options.minLexicalScore,
        minTokenOverlapRatio: options.minTokenOverlapRatio
      });
      return payload.results ?? payload;
    };
  }
}

function normalizeLimit(value: unknown, minValue: number, maxValue: number, fallback: number): number {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) {
    return fallback;
  }
  return Math.max(minValue, Math.min(maxValue, Math.trunc(parsed)));
}

function clampFloat(value: unknown, minValue: number, maxValue: number, fallback: number): number {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) {
    return fallback;
  }
  return Math.max(minValue, Math.min(maxValue, parsed));
}

function normalizePolicyMode(value: unknown): MCPContextPolicyMode {
  const normalized = String(value ?? "").trim().toLowerCase();
  if (normalized === "off" || normalized === "advisory" || normalized === "enforced") {
    return normalized;
  }
  return "advisory";
}

function normalizePolicyProfile(value: unknown): MCPContextPolicyProfileName | undefined {
  const normalized = String(value ?? "").trim().toLowerCase();
  if (!normalized) {
    return undefined;
  }
  if (normalized in MCP_CONTEXT_POLICY_PROFILES) {
    return normalized as MCPContextPolicyProfileName;
  }
  throw new Error(
    `unsupported context policy profile \`${String(value)}\` (allowed: ${Object.keys(MCP_CONTEXT_POLICY_PROFILES).join(", ")})`
  );
}

function resolveContextPolicyProfile(value: unknown): ResolvedContextPolicyProfile {
  const profile = normalizePolicyProfile(value);
  if (!profile) {
    return {};
  }
  const config = MCP_CONTEXT_POLICY_PROFILES[profile];
  return {
    contextPolicyProfile: profile,
    contextPolicyMode: config.contextPolicyMode,
    minRetrievalConfidence: config.minRetrievalConfidence,
    minTotalScore: config.minTotalScore,
    minLexicalScore: config.minLexicalScore,
    minTokenOverlapRatio: config.minTokenOverlapRatio
  };
}

function toArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === "object" && !Array.isArray(value);
}

function coercePayload(response: unknown): unknown {
  if (isRecord(response)) {
    if (isRecord(response.structuredContent)) {
      return response.structuredContent;
    }
    if (isRecord(response.result)) {
      return response.result;
    }
    return response;
  }
  if (Array.isArray(response)) {
    return response;
  }
  return { raw: response };
}

function coerceContextPolicyPayload(
  payload: Record<string, unknown>,
  mode: "off" | "advisory" | "enforced"
): Record<string, unknown> {
  const explainability = isRecord(payload.explainability) ? payload.explainability : null;
  if (explainability && isRecord(explainability.context_policy)) {
    return explainability.context_policy;
  }
  if (isRecord(payload.context_policy)) {
    return payload.context_policy;
  }
  return { mode };
}

function filterContextResults(
  results: unknown[],
  options: { mode: MCPContextPolicyMode; minRetrievalConfidence?: number }
): { results: unknown[]; filteredOut: number } {
  if (options.mode !== "enforced") {
    return { results, filteredOut: 0 };
  }
  const threshold =
    options.minRetrievalConfidence == null ? undefined : clampFloat(options.minRetrievalConfidence, 0, 1, 0.45);
  const kept: unknown[] = [];
  let filteredOut = 0;
  for (const item of results) {
    if (!isRecord(item)) {
      kept.push(item);
      continue;
    }
    if (isRecord(item.context_policy) && "eligible" in item.context_policy) {
      if (Boolean(item.context_policy.eligible)) {
        kept.push(item);
      } else {
        filteredOut += 1;
      }
      continue;
    }
    if (threshold != null && typeof item.retrieval_confidence === "number" && item.retrieval_confidence < threshold) {
      filteredOut += 1;
      continue;
    }
    kept.push(item);
  }
  return { results: kept, filteredOut };
}
